x-pack/plugins/cases/public/components/markdown_editor/use_markdown_session_storage.tsx

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useRef, useState } from 'react';
+import { isEmpty } from 'lodash';
+import useDebounce from 'react-use/lib/useDebounce';
+import useSessionStorage from 'react-use/lib/useSessionStorage';
+import type { FieldHook } from '@kbn/es-ui-shared-plugin/static/forms/hook_form_lib';
+
+const STORAGE_DEBOUNCE_TIME = 500;
+
+export interface SessionStorageType {
+  field: FieldHook<string>;
+  sessionKey: string;
+  initialValue: string | undefined;
+}
+
+export const useMarkdownSessionStorage = ({
+  field,
+  sessionKey,
+  initialValue,
+}: SessionStorageType) => {
+  const [hasConflicts, setHasConflicts] = useState<boolean>(false);
+  const isFirstRender = useRef(true);
+  const initialValueRef = useRef(initialValue);
+
+  const [sessionValue, setSessionValue] = useSessionStorage(sessionKey, '', true);
+
+  if (!isEmpty(sessionValue) && isFirstRender.current) {
+    isFirstRender.current = false;
+    field.setValue(sessionValue);
+  }
+
+  if (initialValue !== initialValueRef.current && initialValue !== field.value) {
+    initialValueRef.current = initialValue;
+    setHasConflicts(true);
+  }
+
+  useDebounce(
+    () => {
+      if (isFirstRender.current) {
+        if (isEmpty(sessionValue) && !isEmpty(field.value)) {
+          /* this condition is used to for lens draft comment, 
+            when user selects and visualization and comes back to Markdown editor,
+            it is a first render for Markdown editor, however field has value of visualization which is not stored in session 
+            hence saving this item in session storage
+          */
+
+          setSessionValue(field.value);
+        }
+        isFirstRender.current = false;
+        return;
+      }
+      setSessionValue(field.value);
```

Comment:
I am not 100% sure and maybe I miss something but I think this can be simplified as:

```suggestion
if (isFirstRender.current) {
        isFirstRender.current = false;
      }

      setSessionValue(field.value);
```

Reply:
I just saw this test `should not update the session value when it is first render` so probably I am wrong 🙂.

Reply:
well, it passed the test and works fine locally as well

Reply:
Weird, with my code the test should not pass as it will update the value in the first render. Could you please remind me why we do not want to update the session value on the first render? I do not remember.